storage/subtree_cache.go

```diff
@@ -0,0 +1,168 @@
+	}
+	prefixSplit := (id.PrefixLenBits - 1) / strataDepth
+	suffixEnd := (id.PrefixLenBits-1)/8 + 1
+	s := Suffix{
+		bits: byte((id.PrefixLenBits-1)%strataDepth) + 1,
+		path: make([]byte, suffixEnd-prefixSplit),
+	}
+	// TODO(al): is all this copying actually necessary?
+	copy(s.path, id.Path[prefixSplit:suffixEnd])
+	if id.PrefixLenBits%8 != 0 {
+		suffixMask := (byte(0x1<<uint((id.PrefixLenBits%8))) - 1) << uint(8-id.PrefixLenBits%8)
+		s.path[len(s.path)-1] &= suffixMask
+	}
+
+	// TODO(al): is all this copying actually necessary?
+	r := make([]byte, prefixSplit)
+	copy(r, id.Path[:prefixSplit])
+	return r, s
+}
+
+// GetNodeHash retrieves the previously written hash and corresponding tree
+// revision for the given node ID.
+func (s *SubtreeCache) GetNodeHash(id NodeID, getSubtree GetSubtreeFunc) (trillian.Hash, int64, error) {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+
+	px, sx := splitNodeID(id)
+	prefixKey := string(px)
+	c := s.subtrees[prefixKey]
+	if c == nil {
+		subID := id
+		subID.PrefixLenBits = len(px) * 8
+		var err error
+		c, err = getSubtree(subID)
+		if err != nil {
+			return nil, -1, err
+		}
+		if c == nil {
+			// storage didn't have one for us, so we'll store an empty proto here
+			// incase we try to update it later on (we won't flush it back to
+			// storage if it's empty.)
+			c = &SubtreeProto{
+				Prefix: px,
+				Depth:  strataDepth,
+				Nodes:  make(map[string]*HashAndRevision),
+			}
+		}
+		if c.Prefix == nil {
+			panic(fmt.Errorf("GetNodeHash nil prefix on %v for id %v with px %#v", c, id.String(), px))
+		}
+
+		s.subtrees[prefixKey] = c
+	}
+
+	nh := c.Nodes[sx.serialize()]
+	if nh == nil {
+		return nil, -1, nil
+	}
+	return nh.Hash, nh.Revision, nil
+}
+
+// SetNodeHash sets a node hash in the cache.
+func (s *SubtreeCache) SetNodeHash(id NodeID, rev int64, h trillian.Hash) error {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	px, sx := splitNodeID(id)
+	prefixKey := string(px)
+	c := s.subtrees[prefixKey]
+	if c == nil {
+		// TODO(al): This is ok, IFF *all* leaves in the subtree are being set,
+		// verify that this is the case when it happens.
+		c = &SubtreeProto{
+			Prefix: px,
+			Nodes:  make(map[string]*HashAndRevision),
+		}
+		s.subtrees[prefixKey] = c
+	}
+	if c.Prefix == nil {
+		panic(fmt.Errorf("nil prefix for %v (key %v)", id.String(), prefixKey))
+	}
+	s.dirtyPrefixes[prefixKey] = true
+	c.Nodes[sx.serialize()] = &HashAndRevision{h, rev}
+	return nil
+}
+
+// Flush causes the cache to write all dirty Subtrees back to storage.
+func (s *SubtreeCache) Flush(setSubtree SetSubtreeFunc) error {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+
+	for k, v := range s.subtrees {
+		if s.dirtyPrefixes[k] {
+			bk := []byte(k)
+			if !bytes.Equal(bk, v.Prefix) {
+				return fmt.Errorf("inconsistent cache: prefix key is %v, but cached object claims %v", bk, v.Prefix)
+			}
+			if len(v.Nodes) > 0 {
+				if err := setSubtree(v); err != nil {
+					return err
```

Comment:
Yes and no :) ...

You *should* be writing in the "future" here (i.e. at `LatestTreeRoot.treeRevision+1`), so partially succeeding updates haven't screwed the currently serving tree, but you have a problem if you want the tree to ever advance.

One way to deal with this is to ensure that any future attempts to write at this n+1th revision contain the same node updates as the set which were attempted in the current update, so e.g. we could have the Log Sequencer/Signer commit to a particular batches of sequenced entries for each tree revision and they will be retried (in order) on any failures.

Similarly for the Log-based Map, the mapping function can batch updates and commit to those, with the same in-order retry constraints.

You could imagine the Log Seq/Signer and Mapper not actually interacting with the trees directly at all, and just writing out work batches into a queue which some internal mechanism reads and applies (with retry) to the underlying trees.  A lighter-weight alternative might be something where the first attempt to write to the future revision stores some hash over the whole batch (before attempting to apply the batch) and any future retries must have the same hash or it's an error - this might make more sense it it turns out it's easy for the Seq/Signer Mapper to exactly replay previous failed updates for example.